larsim/SimFilters/FilterSimPhotonLiteTime_module.cc

```diff
@@ -0,0 +1,170 @@
+
+  std::string const fSimPhotonsLiteCollectionLabel; //!< Label for the sim::SimPhotonsLite data product
+  std::vector<std::vector<int>> const fTimeWindows; //!< Time windows used for filtering. Units are the same as in the sim::SimPhotonsLite
+  int const fMinTotalPhotons; //!< Minimum number of photons inside a window to pass the filter
+  bool const fDebug; //!< Set to true to print (a lot of) debug information.
+  std::size_t const fN; //!< Number of time winows.
+  bool fUseReflectedPhotons; //!< Whether to include reflected photons in the filter.
+  std::string fReflectedLabel; //!< Label for the reflected photons -- "Reflected" by default.
+
+  void CheckTimeWindows() const;
+};
+
+simfilter::FilterSimPhotonLiteTime::FilterSimPhotonLiteTime(
+  fhicl::ParameterSet const& p,
+  art::ProcessingFrame const&)
+  : SharedFilter{p}
+  , fSimPhotonsLiteCollectionLabel(p.get<std::string>("SimPhotonsLiteCollectionLabel"))
+  , fTimeWindows(p.get<std::vector<std::vector<int>>>("TimeWindows"))
+  , fMinTotalPhotons(p.get<int>("MinTotalPhotons"))
+  , fDebug(p.get<bool>("Debug", false))
+  , fN(fTimeWindows.size())
+  , fUseReflectedPhotons(p.get<bool>("UseReflectedPhotons", false))
+  , fReflectedLabel(p.get<std::string>("fReflectedLabel", "Reflected"))
+{
+  CheckTimeWindows();
+
+  // For printing out debug messages, we want to serialize the
+  // event-level calls so that the messages are not garbled.
+  // Otherwise, this module works well for asynchronous event-level
+  // calls.
+  if (fDebug) {
+    serialize();
+  } else {
+    async<art::InEvent>();
+  }
+}
+
+void
+simfilter::FilterSimPhotonLiteTime::CheckTimeWindows() const
+{
+
+  if (fDebug)
+    std::cout << "\tFilterSimPhotonLiteTime: TimeWindows size is "
+              << fTimeWindows.size() << std::endl;
+
+  for (auto const& tw : fTimeWindows) {
+    if (tw.size() != 2)
+      throw cet::exception("FilterSimPhotonLiteTime::CheckTimeWindows")
+        << "Bad time window initialization: time window has wrong size (not 2)."
+        << std::endl;
+
+    if (fDebug)
+      std::cout << "\t\tTimeWindow "
+                << "[" << tw[0] << "," << tw[1] << "]" << std::endl;
+
+    if (tw[0] > tw[1])
+      throw cet::exception("FilterSimPhotonLiteTime::CheckTimeWindows")
+        << "Bad time window initialization: tw[0]>tw[1]. Reverse the order!"
+        << std::endl;
+  }
+}
+
+bool
+simfilter::FilterSimPhotonLiteTime::filter(art::Event& e,
+                                       art::ProcessingFrame const&)
+{
+  auto const& simPhotonsLiteCollection =
+    *e.getValidHandle<std::vector<sim::SimPhotonsLite>>(fSimPhotonsLiteCollectionLabel);
+
+  std::vector<int> sumNPhotonArray(fN, 0);
+
+  const std::vector<sim::SimPhotonsLite> &simPhotonsLiteCollectionReflected = fUseReflectedPhotons ?
+    *e.getValidHandle<std::vector<sim::SimPhotonsLite>>({fSimPhotonsLiteCollectionLabel, fReflectedLabel}) : std::vector<sim::SimPhotonsLite>();
```

Comment:
Just curious, is grabbing `std::vector<sim::SimPhotonsLite> &` instead of `std::vector<art::Ptr<sim::SimPhotonsLite>>` a thread-safety thing?

Reply:
@absolution1, a few things:
- The data product is `std::vector<sim::SimPhotonsLite>`, not `std::vector<art::Ptr<sim::SimPhotonsLite>>`.  In general, `art::Ptr`s should only be used when needing to persist a product reference to disk.
- If the `simPhotonsLiteCollectionReflected` variable were not a reference, then a potentially expensive copy operation would occur by calling `std::vector`'s copy constructor.
- Thread-safety is not an issue here as the framework guarantees that `getValidHandle` is thread-safe, and the `simPhotonsLiteCollectionReflected` variable is immutable (by being marked `const`)

Reply:
Hi @knoepfel 
Thanks for the info.
I was specifically asking why art::Ptr wasn't used, as that would avoid having to use a reference at all.  Using the vector filled with `art::Ptr<PhotonLine>` also wouldn't have the same hefty copy-cost, and it would mean storing a reference to a vector could be avoided.
Aren't the `sim::SimPhotonsLite` coming from disk?  If so, I'm not sure I understand why the `art::Ptr` isn't available to use.

At least in SBND's case, the `sim::SimPhotonsLite` are stored in the same way as the other data products, which are regularly used via `art::Ptrs`:

G4.......... | largeant...... | ..................... | std::vector<simb::MCParticle>........................................ | ..11
G4.......... | largeant...... | ..................... | std::vector<sim::AuxDetSimChannel>................................... | 2272
G4.......... | mcreco........ | ..................... | std::vector<sim::MCTrack>............................................ | ...1
G4.......... | largeant...... | Reflected............ | std::vector<sim::OpDetBacktrackerRecord>............................. | .193
G4.......... | largeant...... | ..................... | art::Assns<simb::MCTruth,simb::MCParticle,sim::GeneratedParticleInfo> | ..11
G4.......... | largeant...... | Reflected............ | std::vector<sim::SimPhotonsLite>..................................... | .193
G4.......... | largeant...... | ..................... | std::vector<sim::SimChannel>......................................... | .680
G4.......... | largeant...... | ..................... | std::vector<sim::SimPhotonsLite>..................................... | .202

Reply:
@absolution1, there appear to be a few misunderstandings:

- Because `art::Ptr`s are designed for persistency, they are three times bigger than a regular pointer.  Large vectors of `art::Ptr`s are thus expensive in terms of memory and I/O.  There has been an over-reliance on them in LArSoft-based code, and we'd like to clean that up.
- The `std::vector<sim::SimPhotonsLite>` type is the type of the data product, either in memory or on disk.  There is no seamless translation from that to `std::vector<art::Ptr<sim::SimPhotonsLite>>`.  Bringing about such a translation requires creating `art::Ptr`s, which in itself is expensive, and does not happen automatically.
- Even if the data product in question could be translated to `std::vector<art::Ptr<sim::SimPhotonsLite>>`, that has nothing to do with whether `simPhotonsLiteCollectionReflected` is an object or a reference:
    ```.cpp
    std::vector<sim::SimPhotonsLite> const& coll = 
      *e.getValidHandle(...); // obj is a reference to the product, no copy
    std::vector<art::Ptr<SimPhotonsLite>> coll = 
      translate(*e.getValidHandle(...)); // coll is owned by value, increasing memory footprint of job
    ```
It's almost always preferable to own objects by value and not by pointer.

Reply:
@knoepfel 

I'm not sure that there's been so much of a misunderstanding as I just don't think I'm explaining myself very well.  I was not definitely not aware how cumbersome the `art::Ptr` is, though.

If LArSoft optimisation is at the point where art::Ptrs are considered hefty then, fine, I can completely see why one would avoid using them.
However, `art::Ptrs` offer a certain degree of protection which, to the best of my knowledge, a reference to a vector does not.
If any of the `art::Ptrs` in the vector go stale and are subsequently accessed then an exception is thrown and everything is bailed gracefully, right?
If the reference to the vector goes stale and is accessed, there would be a segfault?

For sure I was not aware that the `art::Ptr` was considered so hefty these days.
I'm curious about how much memory and CPU time is lost by doing this instead storing the underlying vector as a reference?

`art::Handle<std::vector<sim::SimPhotonsLite> > photons;
e.getByLabel(label, photons);
for (size_t iPhoton = 0; iPhoton < photons.size(); ++iPhoton){
  const art::Ptr<sim::SimPhotonLite> photon(photons, iPhoton);
//stuff to do with the photon
}`

For your last point, I think I've caused a lot of confusion; I was not in any way suggesting that storing by value vs reference are equivalent in terms of CPU/memory cost. I wasn't even suggesting storing an entire `std::vector<art::Ptr<simb::SimPhotonsLite>>` (one could convert each element of the handle as needed instead using the above snippet).  However, seeing as it's been brought up, I've always thought that constructing the `art::Ptrs` and saving them in a new vector vs taking a reference to the entire vector can be preferable, because of the modest global difference in CPU/memory cost and also the safety reasons I mentioned above.  But, I was basing everything on the assumption that art::Ptrs are cheap to construct/copy/move, which it doesn't sound like they are...

Reply:
@absolution1, you've brought up issues that are worth discussing, but at this point it is a discursion from the PR.  The lifetime of the reference in question is localized to the `produce` function call.  Unless the code is caching the reference across `produce` calls (it's not), there will be no segmentation violation...at least not from using a reference.

The calculation of the overhead for a `vector<art::Ptr<T>>` object is not hard:

| Variable | Memory usage (bytes) | Notes |
| -------- | ------------- | ------- |
| `vector<T> const&` | 8 | Size of pointer |
| `vector<art::Ptr<T>>` | 24 x (N + 1) | N is the number of vector elements, and the extra 1 is due to how `std::vector` is implemented |

Even an empty vector owned by value is more expensive than just the reference to an already-owned object.

We should continue further discussion elsewhere, but in short, the reference is just fine in this context.